### DISCUSSION TITLE
KNI Engine and cross platform

### DISCUSSION THREAD

**Discussion by trinith** [MonoGame/MonoGame#8405] [General]
@nkast

Hey, sorry, I have another question for you about the KNI Engine. This isn't a blocker for me but more a question of how to set up projects to achieve what I want.

So in the past, I've tackled crossplatform in MonoGame by creating a portable library referencing `MonoGame.Framework.Portable` and put all of my common game code in there. For each individual platform, I would create a new project and instantiate the game class in my common library, injecting any platform specific dependencies I might need. 

So for example, let's say I had a game with simple gesture controls that I wanted to deploy to both Windows and Android. I might have a project setup along the lines of the following...

* GameCommon (PCL) - MainGame inheriting from `Microsoft.Xna.Framework.Game`. Constructor takes a dependency, `IInputHandler` and uses that to determine its input logic.
* Game_Android (Android Project) - GameActivity.cs instantiates MainGame, passing it `TouchScreenInputHandler` which implements `IInputHandler`, then calls Run.
* Game_Windows (Windows DX Project) - Program.cs instantiates MainGame, passing it `MouseInputHandler` which implements `IInputHandler`, then calls Run.

In this way all of my common game logic was in the PCL library and I only had to do minimal interactions with the platform. I could also completely avoid interspersing my code with `#if defined <platform>`.

When I did all this it was over 5 years ago and it looks like things have maybe changed? I'm looking to work on another project and, as mentioned in my previous thread, I would like to have deploy to web as an option. As a result, I've been exploring the KNI Engine. My tests of web deploy, once I get the screen stuff sorted, have been working great, so thanks for providing that library! However, I seem to be having trouble replicating the same kind of project structure unless I specifically couple to KNI Engine across all platforms.

It looks like `MonoGame.Framework.Portable` doesn't seem to exist anymore. There appears to be a `MonoGame.Framework.Portable.9000` (authored by nkast) which is deprecated and directs to `nkast.Xna.Framework.Ref`. I tried setting my common project up to use this and for my web game, everything works fine. However, for my PC game, I can only make it work if I'm referencing the KNI Engine libraries. If I reference the vanilla MonoGame libraries I get a `PlatformNotSupported` exception thrown at runtime.

So like I said, I'm not blocked. For now I can just make everything on the KNI Engine. However, I would like to know if there are options that let me write everything based on vanilla MonoGame and only use KNI to deploy to web.

I'm open to any insights you, or anybody else, may have to offer here. Thanks!

**Comment by nkast** [comment]
You can't mix the assemblies from KNI and MonoGame. You need to have seperate project files that each reference the corespoding packages.

The _Ref_ library is no longer needed in the latest version 3.12, and you no longer need the bait&switch trick with PrivateAssets to build your libraries. 
In the common library you reference only the framework libraries.
Then each platform project reference those, the common library, and the platform implementation/backend.

The following is a recent example of Gum Forms, targeting both KNI and MonoGame, with a seperate Common library.
https://github.com/vchelaru/Gum/tree/6f200dfccb72a7bbfbca5696705667eba4c8f232/Samples/GumFormsSample

Or the Infiniminer, a somewhat bigger project,
https://github.com/nkast/Infiniminer/tree/b195b9d1a75af04c5539f09cf54071ab2190920c/source/Infiniminer

**Reply by trinith** [reply]
Sorry, I'm not understanding what's happening here. From what I can tell, at least in the Gum Forms Sample, there's a project file for each platform and a solution file for each platform. The Infiniminer setup is a bit bigger and I haven't fully digested that one yet, but on the surface I see similar.

Is there no longer a way to have a single, common library project that is referenced by each individual platform project in the same solution like I was able to with the old PCL library approach?

**Reply by nkast** [reply]
You can use a single common library for all platforms.

What you can't do, is use a library that was build for KNI, in a MonoGame project and vice versa.
The same goes for FNA, XNA, etc.

**Reply by trinith** [reply]
This is the part I'm having trouble with.

How do I make a single common library for all platforms that is not targetting any specific framework? In the past the solution was to use the PCL library (`MonoGame.Framework.Portable`). Is the only current solution to make a common library project (`.csproj`) for each framework (ie, KNI and MonoGame), or is there a modern PCL approach that can be utilized?

**Reply by tomspilman** [reply]
I agree. PCL was great. I think Microsoft dropped support for it, it was perfect for all our scenarios. 

That said we're moving towards a shared c# assembly that uses different native DLLs to define the platform.  It should provide the same sort of functionality, I think.

**Reply by trinith** [reply]
I was just about to post back but I think you've touched on the point I was going to make.

All of this was under the knowledge that a `.csproj` file references all of the source files it builds. In this scenario, maintaining builds for each framework is a bit of nightmare since any time you add/remove a file, you'd have update the other projects as well. This is where the PCL approach really shined as I could just maintain a single project and then reference it any any platform build I desired.

However, it looks like `.csproj` files no longer reference the source files they build... I'm assuming they just pick up everything in their directory and subfolders? Going over the two links that nkast provided, I see very small project files in contrast to the large ones I'm used to. This makes things a little easier to deal with because really, the project file only targets the framework I want to build the source files for. It's still a little ungainly but not nearly as bad as I was originally thinking.

While not as nice as the PCL approach, I think I can deal with this.

@tomspilman 
> That said we're moving towards a shared c# assembly that uses different native DLLs to define the platform. It should provide the same sort of functionality, I think.

Do you mean this is something coming up in a future release, or that the approach that nkast is demonstrating above (and that I now understand a bit better) is what you mean here?

Thanks for the responses guys. It sounds like at the end of the day, we all blame Microsoft and do our best to adapt ;)
